index.js

```diff
@@ -13,23 +13,55 @@ const isAvailable = options => new Promise((resolve, reject) => {
 	});
 });
 
-const getPort = options => {
-	options = Object.assign({}, options);
+const portCheckSequence = function * (ports) {
+	if (ports) {
+		yield * ports;
+	}
+
+	yield 0; // Fall back to 0 if anything else failed
+};
+
+module.exports = async options => {
+	let ports = null;
 
-	if (typeof options.port === 'number') {
-		options.port = [options.port];
+	if (options) {
+		ports = (typeof options.port === 'number') ? [options.port] : options.port;
 	}
 
-	return (options.port || []).reduce(
-		(seq, port) => seq.catch(
-			() => isAvailable(Object.assign({}, options, {port}))
-		),
-		Promise.reject()
-	);
+	for (const port of portCheckSequence(ports)) {
+		const gotPort = await getAvailablePort({...options, port}); // eslint-disable-line no-await-in-loop
+		if (gotPort !== null) {
+			return gotPort;
+		}
+	}
+
+	throw new Error('no available ports found');
 };
 
-module.exports = options => options ?
-	getPort(options).catch(() => getPort(Object.assign(options, {port: 0}))) :
```

Comment:
By removing this logic, the error thrown after the `for` loop is not being handled anymore.

Reply:
Because it shouldn't be handled. It's a real error.

`get-port` got no interface to tell port wasn't found, because it thinks there is no way it can't be found. But it could actually happen, so we're throwing error.

Reply:
You're right about the absence of an interface to indicate a port was not found. However, the issue is that you removed functionality from the current implementation. If there is not any port available in a given array of port numbers, your implementation is discarding a final attempt to get a random port as intended. This behavior should be restored to avoid introducing a breaking change.

`portCheckSequence` is only handling the scenario where `options` is falsy.

Can you confirm this?

Reply:
No, I can't confirm this.
```js
yield 0; // Fall back to 0 if anything else failed
```

Reply:
Yep 👌 My bad here!